### PR TITLE
ARROW-7681: [Rust] Explicitly seeking a BufReader will discard the internal buffer (2)

### PR DESCRIPTION
**Pull request by rdettai**
A fix to 7681 that does not use nightly (as oposed to https://github.com/apache/arrow/pull/6280).

### CODE REVIEW

**Comment by rdettai** [comment]
I would like to add a test on a parquet file with many lines because my first commit passed all tests but was buggy... Where should I add this file ?

**Comment by paddyhoran** [comment]
> Where should I add this file ?

The short answer is [here](https://github.com/apache/arrow-testing) but you should mention it on the mailing list first.  There is reluctance to adding binary data I believe.  The preference is to make progress on the writer implementation so that any test files can be generated.  There has been some initial progress on this [here](https://github.com/apache/arrow/pull/6785).

If there is reluctance to add the file you can comment out the test and open up a JIRA to follow up once the writer is implemented I guess.

**Comment by paddyhoran** [comment]
Changes look reasonable to me BTW, but they need to be reviewed by @sunchao or @sadikovi in general as they are the experts.

File: rust/parquet/src/util/io.rs [hunk 1]
```diff
@@ -31,47 +33,83 @@ pub trait Position {
...
+/// Designed after `arrow::io::RandomAccessFile` and `std::io::BufReader`
 pub struct FileSource<R: ParquetReader> {
-    reader: Mutex<BufReader<R>>,
-    start: u64, // start position in a file
-    end: u64,   // end position in a file
+    reader: RefCell<R>,
+    start: u64,     // start position in a file
+    end: u64,       // end position in a file
+    buf: Vec<u8>,   // the internal buffer `buf` in BufReader
+    buf_pos: usize, // equivalent to the `pos` param in BufReader
```
Comment (sunchao) [inline rust/parquet/src/util/io.rs:48]: This is not super useful as it requires ppl to jump into `BufReader` to check the documentation - maybe add additional comments on what these two fields are for?

File: rust/parquet/src/util/io.rs [hunk 2]
```diff
@@ -31,47 +33,83 @@ pub trait Position {
...
-            self.start += bytes_read as u64;
+        // If we don't have any buffered data and we're doing a massive read
+        // (larger than our internal buffer), bypass our internal buffer
+        // entirely.
+        if self.buf_pos == self.buf_cap && buf.len() >= self.buf.len() {
+            // discard buffer
+            self.buf_pos = 0;
+            self.buf_cap = 0;
+            // read directly into param buffer
+            let mut reader = self.reader.borrow_mut();
```
Comment (sunchao) [inline rust/parquet/src/util/io.rs:96]: nit: maybe we can extract this into a util function?

**Review by sunchao — APPROVED** [review]
Sorry just noticed this! LGTM with a few nits

**Comment by rdettai** [comment]
I'll work on your comments today.

What about the problem we are trying to fix here? Do you agree with the benefits of this fix ?

Also, I'm not sure why a `Mutex` was used here. To we agree that `Refcell` is better as the whole reader is not thread safe, right ?

**Comment by sunchao** [comment]
Yes I think it is beneficial to avoid dropping buffers with `seek`, although it will be nice if the `seek_relative` will be stabilized soon so we can just use that.

> Also, I'm not sure why a Mutex was used here. Do we agree that Refcell is better as the whole reader is not thread safe, right ?

Originally we designed it this way so that we can concurrently read multiple column chunks after obtaining file handle from a single row group.  Since the file handle is shared between these we wanted to provide thread safety on top of it.

**Comment by rdettai** [comment]
> Originally we designed it this way so that we can concurrently read multiple column chunks after obtaining file handle from a single row group. Since the file handle is shared between these we wanted to provide thread safety on top of it.

I'm not sure to understand how this could be the concern of the `FileSource`. Its the reader (file handle) that is passed to it that should be thread safe, and he should not create the Mutex himself.

**Comment by sunchao** [comment]
>  It's the reader (file handle) that is passed to it that should be thread safe

Is [file](https://doc.rust-lang.org/std/fs/struct.File.html) thread-safe? it's not obvious when reading the doc. Plus, here type parameter`R` can be anything that implements `ParquetReader` (with read, seek, length and try_clone capabilities) so you cannot assume all of them guarantees thread-safety.

**Comment by rdettai** [comment]
I completely agree ! What I am saying is that the layer that makes the handle thread safe cannot be the `FileSource` which is in charge of tracking the reading position of one specific reading thread.

But this question of multi-threading seems to be a whole other concern :-) I've made the _nit_ changes you've proposed and improved the tests.

**Comment by sunchao** [comment]
Merged. Thanks @rdettai !

**Comment by jorisvandenbossche** [comment]
It seems that linting for rust was failing here? (and now on master)

**Comment by nevi-me** [comment]
> It seems that linting for rust was failing here? (and now on master)

I'll fix it shortly @jorisvandenbossche

**Comment by sunchao** [comment]
Oops. Sorry my bad. Thanks @nevi-me !

**Comment by rdettai** [comment]
My bad, I thought it was an issue with the linting CI !

**Pull request closed without merging**